libsawtooth/src/client/rest.rs

```diff
@@ -0,0 +1,140 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A client for interacting with sawtooth services.
+
+use reqwest::blocking::Client;
+use serde::Deserialize;
+use std::collections::VecDeque;
+
+use crate::client::Batch as ClientBatch;
+use crate::client::SawtoothClient;
+
+pub use super::error::SawtoothClientError;
+
+/// A client that can be used to interact with sawtooth services. This client handles
+/// communication with the REST API.
+pub struct RestApiSawtoothClient {
+    url: String,
+}
+
+impl RestApiSawtoothClient {
+    /// Create a new 'RestApiSawtoothClient' with the given base 'url'. The URL should be the bind
+    /// endpoint of the sawtooth REST API.
+    pub fn new(url: &str) -> Self {
+        Self { url: url.into() }
+    }
+}
+
+impl SawtoothClient for RestApiSawtoothClient {
+    /// Get all existing batches in the current blockchain.
+    fn list_batches(&self) -> Result<Box<dyn Iterator<Item = ClientBatch>>, SawtoothClientError> {
+        let url = format!("{}/batches", &self.url);
+
+        let batches = Box::new(BatchIter::new(&url).unwrap().map(|batch| ClientBatch {
+            signer_public_key: batch.header.signer_public_key,
+            header_signature: batch.header_signature,
+            txns: batch.transactions.len(),
+        }));
+
+        Ok(batches)
+    }
+}
+
+/// Iterator used for parsing and deserializing batches.
+struct BatchIter {
+    next: Option<String>,
+    cache: VecDeque<Batch>,
+}
+
+impl BatchIter {
+    /// Create a new 'BatchIter' which will make a call to the REST API and load the initial
+    /// cache with the first page of batches.
+    fn new(url: &str) -> Result<Self, SawtoothClientError> {
+        let mut new_iter = Self {
+            next: Some(url.to_string()),
+            cache: VecDeque::with_capacity(0),
+        };
+        new_iter.reload_cache()?;
+        Ok(new_iter)
+    }
+
+    /// If another page of batches exisits, use the 'next' url from the current page and
+    /// reload the cache with the next page of batches.
+    fn reload_cache(&mut self) -> Result<(), SawtoothClientError> {
+        if let Some(url) = &self.next {
+            let request = Client::new().get(url);
+            let response = request.send().map_err(|err| {
+                SawtoothClientError::new_with_source("request failed", err.into())
+            })?;
+
+            let page: Page = response.json().map_err(|err| {
+                SawtoothClientError::new_with_source(
+                    "failed to deserialize response body",
+                    err.into(),
+                )
+            })?;
+
+            self.cache = page.data.into();
+
+            self.next = page.next;
+            Ok(())
+        } else {
+            Err(SawtoothClientError::new("invalid Url"))
+        }
+    }
+}
+
+impl Iterator for BatchIter {
+    type Item = Batch;
```

Comment:
We should consider this iterator should return a `Result<Batch, ClientError>`, so that any errors while fetching can be propogated.  The error created below, doesn't go anywhere currently, but should just be the next (and last) iterator value.

Reply:
Fixed